src/main/java/com/artipie/npm/proxy/NpmProxy.java

```diff
@@ -0,0 +1,361 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.npm.proxy;
+
+import com.artipie.asto.Concatenation;
+import com.artipie.asto.Content;
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.fs.RxFile;
+import com.artipie.asto.rx.RxStorage;
+import com.artipie.asto.rx.RxStorageWrapper;
+import com.artipie.npm.proxy.json.PackageJsonTransformer;
+import com.artipie.npm.proxy.model.NpmAsset;
+import com.artipie.npm.proxy.model.NpmPackage;
+import com.jcabi.log.Logger;
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import io.reactivex.MaybeSource;
+import io.reactivex.Single;
+import io.vertx.core.file.OpenOptions;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.RequestOptions;
+import io.vertx.ext.web.client.WebClientOptions;
+import io.vertx.reactivex.core.Vertx;
+import io.vertx.reactivex.core.buffer.Buffer;
+import io.vertx.reactivex.core.file.AsyncFile;
+import io.vertx.reactivex.ext.web.client.HttpResponse;
+import io.vertx.reactivex.ext.web.client.WebClient;
+import io.vertx.reactivex.ext.web.codec.BodyCodec;
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import net.minidev.json.JSONObject;
+import net.minidev.json.JSONValue;
+import net.minidev.json.parser.JSONParser;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * NPM Proxy.
+ * @since 0.1
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public class NpmProxy {
+    /**
+     * Default connection timeout to remote repo (in millis).
+     */
+    private static final int CONNECT_TIMEOUT = 2_000;
+
+    /**
+     * Default request timeout to remote repo (in millis).
+     */
+    private static final int REQUEST_TIMEOUT = 5_000;
+
+    /**
+     * The settings.
+     */
+    private final NpmProxySettings settings;
+
+    /**
+     * The Vertx instance.
+     */
+    private final Vertx vertx;
+
+    /**
+     * The storage.
+     */
+    private final RxStorage storage;
+
+    /**
+     * Web client.
+     */
+    private final WebClient client;
+
+    /**
+     * Ctor.
+     * @param settings Adapter settings
+     * @param vertx Vertx instance
+     * @param storage Adapter storage
+     */
+    public NpmProxy(final NpmProxySettings settings, final Vertx vertx, final Storage storage) {
+        this.settings = settings;
+        this.vertx = vertx;
+        this.storage = new RxStorageWrapper(storage);
+        this.client =  WebClient.create(vertx, NpmProxy.defaultWebClientOptions());
+    }
+
+    /**
+     * Retrieve package metadata.
+     * @param pkg Package name
+     * @return Package metadata (cached or downloaded from remote repository)
+     * @checkstyle ReturnCountCheck (42 lines)
+     */
+    public Maybe<NpmPackage> getPackage(final String pkg) {
```

Comment:
@brastak I think that "get" prefix is redundant in this method name and parameter should be renamed into `name` as it's package's name, not package itself

Reply:
Which alternatives? ) package is Java reserved word. npmPackage is bad name because of all this class NPM-related. pkg is bad name for API.
Parameter will be renamed

Reply:
oh, sorry, my bad, forgot that `package` is reserved. Ok then, not need for corrections